### PR TITLE
allow BindingPattern in FunctionRestParameter

### PR DESCRIPTION
**Pull request by ajafff**
Second part of #6275

Also adds downlevel emit for the destructured rest param.
I don't know if my change emits correct sourcemaps. @weswigham how do you test this stuff?

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Suggestion: can you put the tests in one file? Seems like they all could be there.

> how do you test this stuff?

You can add a comment at the top of the file like

```ts
// @sourcemap: true
```

**Comment by mhegazy** [comment]
@weswigham can you please review and merge

**Comment by ajafff** [comment]
@weswigham can you please review again? I updated the PR according to your comment a while ago.

**Comment by Yuudaari** [comment]
Any news on this? I just ran into my first use case for it; now that there's nice tuple support I have some type aliases for tuples, it'd be nice to replace the function arguments with the existing tuple so it's updated if the tuple definition ever changes. 

This would allow a method to be able to take a tuple or each value of the tuple easily, since if you're passing a tuple you just use the spread operator, whereas if you're passing the individual values it just works as is. Prevents having to construct new tuples with `tuple = <T extends any[]>(...items: T) => items;`

If you're curious, here's my Real World Code ™ 

```ts
type HighlightSelector = [HighlightType, string | number];

// in a class
public register(component: IComponent, selector: HighlightSelector, until = ComponentEvent.Remove) {
	const selectorId = this.getHighlightSelectorId(...selector);
	this.highlightComponents.set(selectorId, component);
	component.once(until, () => this.highlightComponents.delete(selectorId));
}

private getHighlightSelectorId(...[type, selector]: HighlightSelector) {
	return `${type}:${selector}`;
}

private getHighlightComponent([type, selector]: HighlightSelector) {
	if (type === HighlightType.Selector) {
		return Component.get(`${selector}`);
	}
	
	return this.highlightComponents.get(this.getHighlightSelectorId(type, selector));
}
```

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -29969,10 +29969,6 @@ namespace ts {
                         checkGrammarForDisallowedTrailingComma(parameters, Diagnostics.A_rest_parameter_or_binding_pattern_may_not_have_a_trailing_comma);
                     }
 
-                    if (isBindingPattern(parameter.name)) {
```
Comment (sheetalkamat) [inline src/compiler/checker.ts]: Should this error still be there if its a object binding pattern?

**Pull request merged**